### DISCUSSION TITLE
Set migration stage manually / syncing migration state between dev and prod

### DISCUSSION THREAD

**Discussion by jafioti** [diesel-rs/diesel#4360] [Q&A] [answered]
I apoligize if this is a simple question, but I have a project I have been working on on my local machine, adding and running migrations, and now I am ready to deploy. The problem is, my prod db is still 5 migrations behind, but my schema.rs is up to the latest migration. How would you recommend syncing these two? I can't run the reverts because the changes in the DB of the most recent migrations don't exist in prod to be undone.

Is there some way to do `diesel migration revert` but only change the schema.rs file without affecting the db, so I can revert the schema.rs file to where the prod db is and then run all the migrations?

**Comment by weiznich** [comment, marked as answer]
> The problem is, my prod db is still 5 migrations behind, but my schema.rs is up to the latest migration. How would you recommend syncing these two? I can't run the reverts because the changes in the DB of the most recent migrations don't exist in prod to be undone.

The correct way to sync those two states is to apply the missing 5 migrations to your production database before deploying the new version. This can be done via diesel-cli or via the `embed_migration!()` macro and then running migrations on startup. Depending on your uptime requirements you might do some fancy steps here, or just shut down the old applications, apply the migrations and start the new application.

> Is there some way to do diesel migration revert but only change the schema.rs file without affecting the db, so I can revert the schema.rs file to where the prod db is and then run all the migrations?

That's not required, as you can and should take the existing migrations for that.

**Reply by jafioti** [reply]
What does the diesel cli use to determine the current state in migrations? Like when we run the `diesel migration run` and it starts running at a certian point, what does the cli use to determine where to start from?

**Reply by weiznich** [reply]
There is a `__diesel_schema_migrations` created by diesel. This is then used to keep track of which migrations were run in the past and which are missing.

**Reply by jafioti** [reply]
Ok thanks for the reply, makes sense! I know it's a slightly different question, but when I run the migrations I get an error:
```
unknown field `custom_type_derives`, expected one of `file`, `with_docs`, `filter`, `column_sorting`, `schema`, `patch_file`, `import_types`, `generate_missing_sql_type_definitions` for key `print_schema` at line 8 column 1
```
before any migrations are actually applied.

**Reply by jafioti** [reply]
Nevermind, I guess I was running an old version of the diesel cli. I updated it and it worked. Thank you!

**Reply by weiznich** [reply]
Glad that this is resolved, yes that error looks like a older diesel version that doesn't understand some new config fields.
